### PR TITLE
Updated SPI/SPEI dialog to allow multiple stations

### PR DESCRIPTION
**Pull request by dannyparsons**
- dialog now allows multiple stations
- wrapper R functions added to organise input and output data

**To test**
- Dialog is for monthly data only
- Data must be in date order (within stations)
- Data can't have any gaps i.e. missing months

The dialog should get sensible errors if any of these are not met.

@africanmathsinitiative/developers this is ready for review

### CODE REVIEW

**Review by lilyclements — COMMENTED** [review]
There is an issue with filtering and then pressing OK. I used the Ghana two stations data and created some monthly data. I then realised it didn't have all months given for one of the stations. I filtered the data set to years before 1947, applied it as a filter (not as a subset) and pressed OK. I got the following error:

```
Error in self$get_data_objects(data_name)$add_columns_to_data(col_name,  : 
  Length of new column must be divisible by the length of the data frame
```
If I set as a subset there was no problem.

Also other errors which I assume have been there all along:

1. Line 107 can you remove the colon from "Save Model"
2. The dialog throws an error if I set the value in the "shift" nud to be greater than that of the "scale" nud
3. If I uncheck "Omit Missing Values" I get an error saying data cannot contain missing values. I don't understand when this checkbox is useful therefore?

**Comment by dannyparsons** [comment]
Thanks for the review.

On filtering, since the dialog is adding a new column, there is going to be a problem with this. It is something we should address, but probably more generally than this dialog. If Roger is ok with this "feature" for now, then I'd like to leave this for further work.

@lilyclements would it be easier for you to take over this PR and make the changes you suggested? That might be quicker for you?

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 1]
```diff
@@ -1565,3 +1565,78 @@ make_factor <- function(x, ordered = is.ordered(x)) {
     factor(x, levels = as.character(unique(x)), ordered = ordered)
   }
 }
+
+spei_input <- function(data, station, year, month, element) {
+  if (missing(station)) id_cols <- c(year, month) else id_cols <- c(station, year, month)
+  # SPEI package assumes data is ordered so must be sorted
+  data_sort <- data %>% arrange(!!! rlang::syms(id_cols))
```
[the remark this answers is not in the document]
Comment (dannyparsons) [reply instat/static/InstatObject/R/stand_alone_functions.R:1572]: Yes thanks.

**Comment by dannyparsons** [comment]
For shift/scale we could restrict shift to be strictly lower than scale by setting the max of the nud. scale is the most important parameter, I'm not sure how often shift will be used so happy with that or having the error from the package.

**Comment by lilyclements** [comment]
Happy to make these changes. Just wanted to check what the "Omit Missing Values" checkbox does. If I check it, I get an error saying data cannot contain missing values.

**Comment by dannyparsons** [comment]
It's from the climdex package functions. I think the intention is that it allows you to know whether there are missing values in your data by giving an error if there are and you haven't removed them. It's not the same as what this usually means on our dialogs. What it really means is checked: removing missing values, unchecked: give an error if any missing values.
But since we have the default as checked and there is a sensible error, I think it doesn't matter too much about this.

**Pull request merged**